tools/loader/ImageClassifier.cpp

```diff
@@ -245,96 +346,90 @@ static void printTopKPairs(const std::vector<FloatIndexPair> &topKPairs) {
+    // FP32 so that we don't have to special case the printing of the result
+    // for FP16.
+    SMVarT->convertToType(ElemKind::FloatTy);
+  }
+
+  // Print out the inferred image classification.
+  auto H = SMVarT->getHandle<>();
+  llvm::outs() << "Model: " << functionName << "\n";
+  for (unsigned i = 0; i < inputImageFilenames.size(); i++) {
+    Tensor slice = H.extractSlice(i);
+    auto SH = slice.getHandle<>();
+    llvm::outs() << " File: " << inputImageFilenames[i];
+
+    auto topKPairs = getTopKPairs(SH);
+    printTopKPairs(topKPairs);
+  }
+}
+
 int main(int argc, char **argv) {
   Context ctx;
   // The loader verifies/initializes command line parameters, and initializes
   // the ExecutionEngine and Function.
   Loader loader(argc, argv);
 
-  // Load and process the image data into the data Tensor.
-  Tensor data;
-  loadImagesAndPreprocess(inputImageFilenames, &data);
+  const bool streamInputFilenamesMode =
+      inputImageFilenames.size() == 1 && inputImageFilenames.front() == "-";
 
-  // For ONNX graphs with input in NHWC layout, we transpose the data.
-  switch (imageLayout) {
-  case ImageLayout::NCHW:
-    break;
-  case ImageLayout::NHWC:
-    Tensor dataNHWC;
-    data.transpose(&dataNHWC, NCHW2NHWC);
-    data = std::move(dataNHWC);
-    break;
-  }
+  GLOW_ASSERT(!(streamInputFilenamesMode && emittingBundle()) &&
+              "Cannot emit a bundle and also stream inputs.");
 
-  // The image name that the model expects must be passed on the command line.
-  const char *inputName = modelInputName.c_str();
+  // Use to make sure we only compile once.
+  bool isFirstRun = true;
 
-  // Create the model based on the input model format.
-  std::unique_ptr<ProtobufLoader> LD;
-  bool c2Model = !loader.getCaffe2NetDescFilename().empty();
-  if (c2Model) {
-    LD.reset(new Caffe2ModelLoader(
-        loader.getCaffe2NetDescFilename(), loader.getCaffe2NetWeightFilename(),
-        {inputName}, {&data.getType()}, *loader.getFunction()));
-  } else {
-    LD.reset(new ONNXModelLoader(loader.getOnnxModelFilename(), {inputName},
-                                 {&data.getType()}, *loader.getFunction()));
-  }
-
-  // Allocate tensors to back all inputs and outputs.
-  ctx.allocate(loader.getModule()->getPlaceholders());
-  if (convertInAndOutToFp16) {
-    // Convert the raw input to fp16.
-    data.convertToType(ElemKind::Float16Ty);
-    TypeAToTypeBFunctionConverter converter(
-        *loader.getFunction(), ElemKind::FloatTy, ElemKind::Float16Ty);
-    for (auto *placeholder : loader.getModule()->getPlaceholders()) {
-      converter.convertPlaceholder(*placeholder, &ctx);
+  // These will be set during the first run.
+  Placeholder *inputImagePH = nullptr;
+  Tensor *SMVarT = nullptr;
+  do {
+    // If streaming filenames, then load the next set of them.
+    if (streamInputFilenamesMode) {
+      getNextImageFilenames(&inputImageFilenames);
```

Comment:
Shouldn't we have something like `while(getNextImageFilenames)` instead of controlling the loop through `streamInputFilenamesMode`, which will always be true or always be false for a given run. I.e., we don't have a way of terminating the program.

Reply:
I updated it to include something like this (needs to be a little more complicated than `while(getNextImageFilenames)`)